sockeye/inference.py

```diff
@@ -1464,35 +1484,38 @@ def _beam_search(self,
             if self.restrict_lexicon:
                 best_word_indices = vocab_slice_ids.take(best_word_indices)
 
-            # (4) Reorder beam data according to best hypotheses indices (ascending)
-            finished = mx.nd.take(finished, best_hyp_indices)
-            lengths = mx.nd.take(lengths, best_hyp_indices)
-            sequences = mx.nd.take(sequences, best_hyp_indices)
-            attention_scores = mx.nd.take(attention_scores, best_hyp_indices)
-            attentions = mx.nd.take(attentions, best_hyp_indices)
+            # (4) Reorder fixed-size beam data according to best_hyp_indices (ascending)
+            finished, lengths, attention_scores = self._sort_by_index.forward(best_hyp_indices,
+                                                                              finished,
+                                                                              lengths,
+                                                                              attention_scores)
 
             # (5) Normalize the scores of newly finished hypotheses. Note that after this until the
             # next call to topk(), hypotheses may not be in sorted order.
-
-            all_finished = ((best_word_indices == C.PAD_ID) + (best_word_indices == self.vocab_target[C.EOS_SYMBOL]))
-            newly_finished = all_finished - finished
-            scores_accumulated = mx.nd.where(newly_finished,
-                                             scores_accumulated / self.length_penalty(lengths),
-                                             scores_accumulated)
-            finished = all_finished
+            finished, scores_accumulated = self._normalize_finished.forward(best_word_indices,
+                                                                            finished,
+                                                                            scores_accumulated,
+                                                                            lengths)
 
             # (6) Prune out low-probability hypotheses. Pruning works by setting entries `inactive`.
             if self.beam_prune > 0.0:
-                inactive = self.prune(scores_accumulated, finished)
+                inactive = self._prune_hyps.forward(scores_accumulated, finished, self.inf_array)
```

Comment:
why not also pull in the where's below?

Reply:
We just had changed this [recently](https://github.com/awslabs/sockeye/commit/fc440df9cccce7d933b520f320a546cecf82325b) to make the interaction between `inactive` and affected data structures more clear.
I could change it but the trouble is that hybrid where's are more strict w.r.t to shape: thinks like `where((5,), (5,1), (5,)` don't work in HybridBlocks but imperatively apparently. 
I tried for a while to get the shapes of all the auxiliary arrays and `scores_accumulated` etc. to work fine, but its adding more reshapes here and there.

Reply:
will do